drivers/at86rf215/at86rf215_o-qpsk.c

```diff
@@ -256,22 +332,127 @@ void _end_configure_OQPSK(at86rf215_t *dev)
     at86rf215_enable_radio(dev, BB_MROQPSK);
 }
 
-int at86rf215_configure_legacy_OQPSK(at86rf215_t *dev, bool high_rate)
+int at86rf215_configure_OQPSK(at86rf215_t *dev, uint8_t chips, uint8_t mode)
 {
-    /* select 'mode' that would result in the approprate MR-O-QPSK data rate */
-    uint8_t mode  = high_rate ? 3 : 2;
-    uint8_t chips = is_subGHz(dev) ? BB_FCHIP1000 : BB_FCHIP2000;
+    if (chips > BB_FCHIP2000) {
+        DEBUG("[%s] invalid chips: %d\n", __func__, chips);
+        return -EINVAL;
+    }
+
+    if (mode > 4) {
+        DEBUG("[%s] invalid mode: %d\n", __func__, mode);
+        return -EINVAL;
+    }
+
+    /* mode 4 only supports 2000 kchip/s */
+    if (mode == 4 && chips != BB_FCHIP2000) {
+        DEBUG("[%s] mode 4 only supports 2000 kChip/s\n", __func__);
+        return -EINVAL;
+    }
 
     at86rf215_await_state_end(dev, RF_STATE_TX);
 
     /* disable radio */
     at86rf215_reg_write(dev, dev->BBC->RG_PC, 0);
 
-    _set_legacy(dev, high_rate);
-    _set_csma_backoff_period(dev, chips, mode);
+    _set_mode(dev, mode);
+    _set_chips(dev, chips);
+    _set_csma_backoff_period(dev, chips);
     _set_ack_timeout(dev, chips, mode);
 
     _end_configure_OQPSK(dev);
 
     return 0;
 }
+
+int at86rf215_configure_legacy_OQPSK(at86rf215_t *dev, bool high_rate)
+{
+    at86rf215_await_state_end(dev, RF_STATE_TX);
+
+    /* disable radio */
+    at86rf215_reg_write(dev, dev->BBC->RG_PC, 0);
+
+    _set_legacy(dev, high_rate);
+    _set_csma_backoff_period_legacy(dev);
+    _set_ack_timeout_legacy(dev);
+
+    _end_configure_OQPSK(dev);
+
+    return 0;
+}
+
+int at86rf215_OQPSK_set_chips(at86rf215_t *dev, uint8_t chips)
+{
+    uint8_t mode;
+
+    mode = at86rf215_reg_read(dev, dev->BBC->RG_OQPSKPHRTX);
+
+    if (mode & AT86RF215_OQPSK_MODE_LEGACY) {
+        DEBUG("[%s] can't set chip rate in legacy mode\n", __func__);
+        return -1;
+    }
```

Comment:
Aren't there two chips options? 1000 and 2000?

Reply:
Or is it because the proprietary part is not supported?

Reply:
It's always 2000 in 2.4 GHz and 1000 in sub-GHz band - this is fixed.